src/contract/link.rs

```diff
@@ -0,0 +1,233 @@
+                (None, true) => Vec::new(),
+                (Some(ctor), _) => ctor.encode_input(Vec::new(), &params)?,
+            }
+        };
+
+        let libraries = Vec::new();
+
+        Ok(Linker {
+            web3,
+            context,
+            contract_bytecode,
+            encoded_contructor_params,
+            libraries,
+            _instance: PhantomData,
+        })
+    }
+
+    /// Adds a library instance to the linker.
+    pub fn link_library<L>(self, library: L) -> Self
+    where
+        L: LibraryInstance,
+        I: DependsOn<L>,
+    {
+        self.link_library_at(library.name(), library.address())
+    }
+
+    /// Adds a library to the linker by name and address.
+    pub fn link_library_at<S>(self, name: S, address: Address) -> Self
+    where
+        S: AsRef<str>,
+    {
+        self.add_library(name.as_ref(), Library::Resolved(address))
+    }
+
+    /// Adds a library to deploy.
+    pub fn deploy_library<L>(self) -> Self
+    where
+        L: DeployLibrary,
+        I: DependsOn<L>,
+    {
+        self.add_library(L::name(), Library::Pending(L::bytecode()))
+    }
+
+    /// Add a library to the current dependency graph.
+    ///
+    /// Note that this method always succeeds, this is because we do not link
+    /// incrementally, but rather when `build` is called so that linking errors
+    /// only need to be handled in one place.
+    fn add_library(mut self, name: &str, library: Library) -> Self {
+        let name = name.to_owned();
+        self.libraries.push((name, library));
+        self
+    }
+
+    /// Links the libraries and binaries together and returns a verified
+    /// deployment.
+    ///
+    /// This method will return an error if it finds unresolved or unused
+    /// libraries during the linking process.
+    pub fn build(mut self) -> Result<Deployment, LinkerError> {
+        let mut pending_libraries = HashMap::new();
+        for (name, library) in self.libraries {
+            match library {
+                Library::Resolved(address) => self.contract_bytecode.link(&name, address)?,
+                Library::Pending(bytecode) => {
+                    // NOTE: Check that the map doesn't contain the library
+                    //   first because inserting moves `name` into the map.
+                    if pending_libraries.contains_key(&name) {
+                        return Err(LinkerError::UnusedDependency(name));
+                    }
+                    pending_libraries.insert(name, bytecode);
+                }
+            }
+        }
+
+        let mut libraries_to_deploy = Vec::new();
+        for library in self.contract_bytecode.undefined_libraries() {
+            if let Some((name, bytecode)) = pending_libraries.remove_entry(library) {
+                let bytes = match bytecode.try_to_bytes() {
+                    Some(bytes) => bytes,
+                    None => return Err(LinkerError::NestedDependencies(name)),
```

Comment:
So we don't support libraries that need linking themselves?

Reply:
I removed it here: https://github.com/gnosis/ethcontract-rs/pull/144/commits/934c6ed1995593563971cccb00271d3c07a1ccc6 I want to split it into multiple PRs. It will be too much otherwise.